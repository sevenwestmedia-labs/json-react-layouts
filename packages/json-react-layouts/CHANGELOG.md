# json-react-layouts

## 3.0.0-beta.9

### Minor Changes

-   c3d964b: Switched typescript-log to use debug package, see readme for how to enable debug logging

## 3.0.0-beta.8

### Patch Changes

-   de1994c: Fixed crash when using hooks inside useRuntimeParams

## 3.0.0-beta.7

### Patch Changes

-   ddf9a2e: Fixed issue where middleware is violating the rules of hooks

## 3.0.0-beta.6

### Minor Changes

-   9a6621d: Upgraded dependencies

## 3.0.0-beta.5

### Minor Changes

-   885b1fa: Add ability for a composition to provide additional props to all components rendered inside it

    This opens up the possibility of compositions to pass width ratio information down to children, enabling sorta container queries

## 3.0.0-beta.4

### Patch Changes

-   Restore default generic param for DataDefinition type

## 3.0.0-beta.3

### Patch Changes

-   Allow checked composition to be used as a nested composition

## 3.0.0-beta.2

### Patch Changes

-   1207d38: Simplified types further

## 3.0.0-beta.1

### Patch Changes

-   0083a12: Each component and composition must be wrapped in a layout.component() or layout.composition()

    This is because TypeScript does not yet support variadic types, meaning we can't accurately enforce the types on a ...rest style function.

    Before:

    ```
    {
        type: 'test-composition',
        contentAreas: {
            main: [
                {
                    type: 'test-with-data',
                    props: { dataDefinitionArgs: { dataArg: 'Foo' } },
                },
            ],
        },
        props: {},
    }
    ```

    After:

    ```
    layout.composition({
        type: 'test-composition',
        contentAreas: {
            main: [
                layout.component({
                    type: 'test-with-data',
                    props: { dataDefinitionArgs: { dataArg: 'Foo' } },
                }),
            ],
        },
        props: {},
    })
    ```

## 3.0.0-beta.0

### Major Changes

-   Restructured public API slightly to reduce bundle size and type complexity
