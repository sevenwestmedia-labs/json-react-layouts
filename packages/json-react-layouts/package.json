{
    "name": "json-react-layouts",
    "version": "3.0.0",
    "main": "out-tsc/index.js",
    "publishConfig": {
        "main": "dist/cjs/index.js",
        "module": "dist/esm/index.js"
    },
    "sideEffects": false,
    "license": "MIT",
    "repository": "github:sevenwestmedia-labs/json-react-layouts",
    "dependencies": {
        "debug": "^4.0.0"
    }
}
