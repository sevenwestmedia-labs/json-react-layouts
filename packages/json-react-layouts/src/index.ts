export { RenderFunction, ComponentRegistration, ComponentInformation } from './ComponentRegistrar'

export { RendererMiddleware, MiddlwareHandler, MiddlwareServices } from './middlewares'

export {
    LayoutApi,
    RenderLayouts,
    CheckedCompositionInformation,
    CheckedComponentInformation,
} from './LayoutApi'
export {
    CompositionInformation,
    CompositionRegistration,
    CompositionRenderFunction,
    CompositionRenderProps,
    CompositionRegistrations as CompositionRegistrar,
} from './CompositionRegistrar'
export { getRegistrationCreators } from './get-registration-creators'
export {
    LayoutRegistration,
    ComponentCheckedMessage,
    CompositionCheckedMessage,
    NestedCompositionProps,
} from './LayoutRegistration'

export {
    flatMap,
    isNestedComposition,
    getComponentsInCompositions,
    ContentAreaData,
} from './helpers'
