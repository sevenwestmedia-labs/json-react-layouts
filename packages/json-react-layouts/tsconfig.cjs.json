{
    "extends": "../../tsconfig.settings.json",
    "compilerOptions": {
        "outDir": "./dist/cjs" /* Redirect output structure to the directory. */,
        "rootDir": "./src" /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
    },
    "exclude": ["dist"]
}
