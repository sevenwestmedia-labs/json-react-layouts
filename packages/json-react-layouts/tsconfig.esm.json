{
    "extends": "../../tsconfig.settings.json",
    "compilerOptions": {
        "module": "ES2015",
        "outDir": "./dist/esm" /* Redirect output structure to the directory. */,
        "rootDir": "./src" /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
    },
    "exclude": ["dist"]
}
