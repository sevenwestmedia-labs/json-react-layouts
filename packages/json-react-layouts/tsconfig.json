{
    "extends": "../../tsconfig.settings.json",
    "compilerOptions": {
        "outDir": "./out-tsc",
        "rootDir": "./src"
    },
    "exclude": ["dist", "out-tsc"]
}
