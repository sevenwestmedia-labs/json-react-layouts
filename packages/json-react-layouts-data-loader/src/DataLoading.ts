import { LayoutApi } from 'json-react-layouts'

export type LoadData<DataLoadArguments extends {}, TData, Services extends {}> = (
    dataDefinitionArgs: DataLoadArguments,
    services: Services,
    context: { resourceType: string; paramsCacheKey: string },
) => Promise<TData>

export interface DataDefinition<
    DataLoadArguments extends {},
    TData,
    Services extends {},
    AdditionalParams extends {} = {}
> {
    /** Custom React Hook to provide additional dynamic parameters to the data loader */
    useRuntimeParams?: (
        dataDefinitionArgs: DataLoadArguments,
        services: Services,
    ) => AdditionalParams
    loadData: LoadData<DataLoadArguments & AdditionalParams, TData, Services>
}

export type MaybeLoaded<TData> = { loaded: false } | { loaded: true; result: TData }
export interface ComponentState<TData> {
    data: MaybeLoaded<TData>
}

export interface LoadArguments<Services extends object> {
    dataDefinition: DataDefinition<any, any, Services, any>
    dataDefinitionArgs: any
    layout: LayoutApi<any, any, any, any, Services>
}
