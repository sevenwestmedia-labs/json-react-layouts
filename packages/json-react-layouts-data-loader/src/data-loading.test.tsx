import React from 'react'
import { act } from 'react-dom/test-utils'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import { init } from '.'
import { DataDefinition } from './DataLoading'
import { DataLoaderResources, DataProvider } from 'react-ssr-data-loader'
import { getRegistrationCreators, LayoutRegistration } from 'json-react-layouts'
import { mount, configure } from 'enzyme'

configure({ adapter: new Adapter() })

it('can load data for component', async () => {
    const resources = new DataLoaderResources<{}>()
    const { middleware, createRegisterableComponentWithData } = init<{}>(resources)

    const testComponentWithDataRegistration = createRegisterableComponentWithData(
        'test-with-data',
        lengthCalculatorDataDefinition,
        (props, data) => {
            return (
                <TestComponentWithData
                    length={data.loaded ? data.result : undefined}
                    {...props}
                    {...{ dataProps: { data } }}
                />
            )
        },
    )

    const layout = LayoutRegistration()
        .registerComponents(registrar =>
            registrar
                .registerComponent(testComponentWithDataRegistration)
                .registerMiddleware(middleware),
        )
        .registerCompositions(registrar =>
            registrar.registerComposition(testCompositionRegistration),
        )
    const renderers = layout.createRenderers({
        services: {},
    })

    const wrapper = mount(
        <DataProvider resources={resources} globalProps={{}}>
            {renderers.renderCompositions(
                layout.composition({
                    type: 'test-composition',
                    contentAreas: {
                        main: [
                            layout.component({
                                type: 'test-with-data',
                                props: { dataDefinitionArgs: { dataArg: 'Foo' } },
                            }),
                        ],
                    },
                    props: {},
                }),
            )}
        </DataProvider>,
    )

    expect(wrapper.find(TestComponentWithData).text()).toBe('Loading')
    await act(() => new Promise(resolve => setTimeout(resolve)))

    const component = wrapper.update().find(TestComponentWithData)
    expect(component.text()).toBe('Length: 3')
    expect(component.props()).toMatchSnapshot()
})

it('cap wrap data load function', async () => {
    let wrapArgs: any
    let wrapServices: any
    let wrapContext: any
    const resources = new DataLoaderResources<{ serviceProp: 'example' }>()
    const { middleware, createRegisterableComponentWithData } = init<{ serviceProp: 'example' }>(
        resources,
        load => (args, services, context) => {
            wrapArgs = args
            wrapServices = services
            wrapContext = context
            return load(args, services, context)
        },
    )

    const testComponentWithDataRegistration = createRegisterableComponentWithData(
        'test-with-data',
        lengthCalculatorDataDefinition,
        (props, data) => {
            return (
                <TestComponentWithData
                    length={data.loaded ? data.result : undefined}
                    {...props}
                    {...{ dataProps: { data } }}
                />
            )
        },
    )

    const layout = LayoutRegistration<{ serviceProp: 'example' }>()
        .registerComponents(registrar =>
            registrar
                .registerComponent(testComponentWithDataRegistration)
                .registerMiddleware(middleware),
        )
        .registerCompositions(registrar =>
            registrar.registerComposition(testCompositionRegistration),
        )

    const renderers = layout.createRenderers({
        services: { serviceProp: 'example' },
    })

    mount(
        <DataProvider resources={resources} globalProps={{ serviceProp: 'example' }}>
            {renderers.renderCompositions(
                layout.composition({
                    type: 'test-composition',
                    contentAreas: {
                        main: [
                            layout.component({
                                type: 'test-with-data',
                                props: { dataDefinitionArgs: { dataArg: 'Foo' } },
                            }),
                        ],
                    },
                    props: {},
                }),
            )}
        </DataProvider>,
    )

    expect(wrapArgs).toEqual({ dataArg: 'Foo' })
    expect(wrapServices).toEqual({ serviceProp: 'example' })
    expect(wrapContext).toEqual({
        paramsCacheKey: '7c4bd5d4',
        resourceType: 'component-data-loader',
    })
})

it('component can provide additional arguments dynamically', async () => {
    const resources = new DataLoaderResources<{}>()
    const { middleware, createRegisterableComponentWithData } = init<{}>(resources)
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    let updateMultiplier: (multiplier: number) => void = () => {}

    const lengthCalculatorWithMultiplierDataDefinition: DataDefinition<
        { dataArg: string },
        number,
        {},
        { multiplier: number }
    > = {
        // Additional params can come from anywhere, for instance redux or
        // other environmental variables (window.location?)
        useRuntimeParams: () => {
            const [multiplier, setMultiplier] = React.useState(2)
            React.useEffect(() => {
                updateMultiplier = setMultiplier
            }, [])

            return {
                multiplier,
            }
        },
        loadData: props =>
            new Promise(resolve =>
                setTimeout(() => {
                    resolve(props.dataArg.length * props.multiplier)
                }),
            ),
    }

    const testComponentWithDataRegistration = createRegisterableComponentWithData(
        'test-with-data',
        lengthCalculatorWithMultiplierDataDefinition,
        (props, data) => {
            return (
                <TestComponentWithData
                    length={data.loaded ? data.result : undefined}
                    {...props}
                    {...{ dataProps: { data } }}
                />
            )
        },
    )

    const layout = LayoutRegistration()
        .registerComponents(registrar =>
            registrar
                .registerComponent(testComponentWithDataRegistration)
                .registerMiddleware(middleware),
        )
        .registerCompositions(registrar =>
            registrar.registerComposition(testCompositionRegistration),
        )

    const renderers = layout.createRenderers({
        services: {},
    })

    const wrapper = mount(
        <DataProvider resources={resources} globalProps={{}}>
            {renderers.renderCompositions(
                layout.composition({
                    type: 'test-composition',
                    contentAreas: {
                        main: [
                            layout.component({
                                type: 'test-with-data',
                                props: { dataDefinitionArgs: { dataArg: 'Foo' } },
                            }),
                        ],
                    },
                    props: {},
                }),
            )}
        </DataProvider>,
    )

    await act(() => new Promise(resolve => setTimeout(resolve)))

    let component = wrapper.update().find(TestComponentWithData)
    expect(component.text()).toBe('Length: 6')
    expect(component.props()).toMatchObject({
        dataProps: {
            data: {
                dataDefinitionArgs: { dataArg: 'Foo', multiplier: 2 },
            },
        },
    })

    act(() => {
        updateMultiplier(3)
    })
    await act(() => new Promise(resolve => setTimeout(resolve)))

    component = wrapper.update().find(TestComponentWithData)
    expect(component.text()).toBe('Length: 9')
    expect(component.props()).toMatchObject({
        dataProps: {
            data: {
                dataDefinitionArgs: { dataArg: 'Foo', multiplier: 3 },
            },
        },
    })
})

const { createRegisterableComposition } = getRegistrationCreators<{}>()

// Test component with data
const TestComponentWithData: React.FC<{ length: number | undefined }> = ({ length }) => (
    <div>{length ? `Length: ${length}` : 'Loading'}</div>
)

const TestComposition: React.FC<{ main: React.ReactElement<any> }> = props => (
    <div>{props.main}</div>
)

const testCompositionRegistration = createRegisterableComposition<'main'>()(
    'test-composition',
    contentAreas => <TestComposition main={contentAreas.main} />,
)

const lengthCalculatorDataDefinition: DataDefinition<{ dataArg: string }, number, {}, {}> = {
    loadData: props =>
        new Promise(resolve =>
            setTimeout(() => {
                resolve(props.dataArg.length)
            }),
        ),
}
