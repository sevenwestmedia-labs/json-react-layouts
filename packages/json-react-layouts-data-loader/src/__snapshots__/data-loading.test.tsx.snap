// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can load data for component 1`] = `
Object {
  "dataProps": Object {
    "data": Object {
      "dataDefinitionArgs": Object {
        "dataArg": "Foo",
      },
      "loaded": true,
      "result": 3,
    },
  },
  "layoutType": "test-with-data",
  "length": 3,
  "renderPath": "[0]test-composition/main/[0]",
}
`;

exports[`can load data for composition 1`] = `
Object {
  "dataProps": Object {
    "data": Object {
      "dataDefinitionArgs": Object {
        "dataArg": "Foo",
      },
      "loaded": true,
      "result": 3,
    },
  },
  "layoutType": "test-with-data",
  "length": 3,
  "main": <ComponentsRenderer
    additionalComponentProps={Object {}}
    componentMiddleware={[Function]}
    componentRegistrations={
      Object {
        "get": [Function],
        "isRegistered": [Function],
      }
    }
    componentRenderPath="[0]test-with-data/main"
    components={
      Array [
        Object {
          "props": Object {},
          "type": "test-component",
        },
      ]
    }
    layoutApi={
      Object {
        "component": [Function],
        "componentRegistrations": Object {
          "get": [Function],
          "isRegistered": [Function],
        },
        "composition": [Function],
        "compositionRegistrations": Object {
          "get": [Function],
        },
        "createRenderers": [Function],
        "nestedComposition": [Function],
      }
    }
    services={Object {}}
  />,
  "renderPath": "[0]test-with-data",
}
`;
